### PR TITLE
Make early lints translatable

### PR DESCRIPTION
**Pull request by Xiretza**
<del>Requires https://github.com/projectfluent/fluent-rs/pull/353.</del> https://github.com/rust-lang/rust/commit/5134a04eaa32b168cf5998a6ec13199356e2e017

r? diagnostics

### CODE REVIEW

**Comment by fmease** [comment]
Are there any blockers left apart from a rebase?

**Comment by Xiretza** [comment]
Yep, I can also patch those `NUMBER()` calls out and put a TODO there instead, at the cost of slightly worse diagnostics until the next fluent release.

**Comment by Xiretza** [comment]
Rebased, and the use case for `NUMBER()` has since disappeared (https://github.com/rust-lang/rust/commit/5134a04eaa32b168cf5998a6ec13199356e2e017), so there are no blockers now.

**Comment by davidtwco** [comment]
Apologies for the delay in my reviewing, had a busy couple weeks, will let @fmease handle this.

**Review by fmease — CHANGES_REQUESTED** [review]
Thanks a lot. this is great work!

I've added a lot of review comments for things that are secondary or preexisting even. You don't need to address them in this PR, they can be fixed in follow-up PRs (by you or my me). Basically drive-by commentary.

I g2g rn unfortunately but I'd like to see `emit_buffered_lint` being made non-generic as noted in one of my comments. You should be able to pick out what is relevant and what isn't. In a few hours, I will be back again and can push it into the merge queue with a high priority

File: compiler/rustc_lint/src/context/diagnostics.rs [hunk 1]
```diff
@@ -353,3 +356,122 @@ pub(super) fn builtin(sess: &Session, diagnostic: BuiltinLintDiag, diag: &mut Di
...
+                be referred to by absolute paths"
+                .into()
+        }
+        BuiltinLintDiag::ElidedLifetimesInPaths(_, _, _, _) => {
+            "hidden lifetime parameters in types are deprecated".into()
+        }
+        BuiltinLintDiag::UnknownCrateTypes(_, _, _) => "invalid `crate_type` value".into(),
+        BuiltinLintDiag::UnusedImports { span_snippets, .. } => format!(
+            "unused import{}{}",
+            pluralize!(span_snippets.len()),
```
[the remark this answers is not in the document]
Comment (Xiretza) [reply compiler/rustc_lint/src/context/diagnostics.rs]: Good catch, I think this was potentially wrong previously: the length is used to pluralize the list of `span_snippets`, so it should be `span_snippets.len()` in any case.

File: compiler/rustc_lint_defs/src/lib.rs [hunk 2]
```diff
@@ -745,16 +781,6 @@ impl LintBuffer {
...
-        &mut self,
-        lint: &'static Lint,
-        id: NodeId,
-        sp: impl Into<MultiSpan>,
-        msg: impl Into<DiagMessage>,
-    ) {
-        self.add_lint(lint, id, sp.into(), BuiltinLintDiag::Normal(msg.into()))
-    }
-
     pub fn buffer_lint_with_diagnostic(
```
[the remark this answers is not in the document]
Comment (Xiretza) [reply compiler/rustc_lint_defs/src/lib.rs:787]: Done.

File: compiler/rustc_session/src/parse.rs [hunk 3]
```diff
@@ -300,23 +300,6 @@ impl ParseSess {
...
-            buffered_lints.push(BufferedEarlyLint {
-                span: span.into(),
-                node_id,
-                lint_id: LintId::of(lint),
-                diagnostic: BuiltinLintDiag::Normal(msg.into()),
-            });
-        });
-    }
-
     pub fn buffer_lint_with_diagnostic(
```
[the remark this answers is not in the document]
Comment (Xiretza) [reply compiler/rustc_session/src/parse.rs:303]: Done.

File: compiler/rustc_expand/src/base.rs [hunk 4]
```diff
@@ -1367,14 +1367,14 @@ fn pretty_printing_compatibility_hack(item: &Item, sess: &Session) -> bool {
...
-                                        ast::CRATE_NODE_ID,
-                                        BuiltinLintDiag::ProcMacroBackCompat(
-                                        "older versions of the `rental` crate will stop compiling in future versions of Rust; \
-                                        please update to `rental` v0.5.6, or switch to one of the `rental` alternatives".to_string()
-                                        )
-                                    );
+                                    PROC_MACRO_BACK_COMPAT,
+                                    item.ident.span,
+                                    ast::CRATE_NODE_ID,
+                                    BuiltinLintDiag::ProcMacroBackCompat {
```
[the remark this answers is not in the document]
Comment (Xiretza) [reply compiler/rustc_expand/src/base.rs:1373]: I wasn't sure if other crates might be added here in the future (there were more in the past), so I went for the possibly overkill solution.

**Review by Xiretza — COMMENTED** [review]
<del>Still need to address the most pressing concern (making `emit_buffered_lint` not generic), but I got most of the rest done.</del>

https://github.com/rust-lang/rust/pull/124417#discussion_r1607169744

**Review by fmease — APPROVED** [review]
Thanks for applying my suggestions! One final rebase necessary, then this is ready to get shipped!

**Comment by Xiretza** [comment]
Alright, `check_cfg` rebased, and even prettified a bit in the process!

**Comment by Xiretza** [comment]
Weird, the push hook didn't catch some bad formatting, hopefully that was the last force-push.

**Comment by fmease** [comment]
> Weird, the push hook didn't catch some bad formatting, hopefully that was the last force-push.

Ah, I've seem some people post about similar experiences on Zulip, no worries.

**Comment by nnethercote** [comment]
Some sizeable perf regressions. Is that expected?

**Comment by fmease** [comment]
No, it's not expected. I think I know what's going wrong. Sending a fix shortly.

**Comment by Xiretza** [comment]
No, not really - would be interesting to know if this is just for lints that are actually emitted or even for ones that get silenced.

**Comment by fmease** [comment]
@Xiretza We no longer decorate the lint `Diag` lazily inside `span_lint_with_diagnostics` but eagerly in `diagnostics::emit_buffered_lint`.

**Comment by fmease** [comment]
In #125410 I was able to partially restore the old performance characteristics. See https://github.com/rust-lang/rust/pull/125410#issuecomment-2125971223. Doing some naive calculations in my head, I guess that's not sufficient. Anyways, Xiretza and/or I can conduct some more perf experiments after #125410 has been merged.

**Pull request merged**